code/addons/test/src/components/TestProviderRender.tsx

```diff
@@ -425,14 +452,25 @@ export const TestProviderRender: FC<
 };
 
 function useConfig(api: API, providerId: string, initialConfig: Config) {
-  const [currentConfig, setConfig] = useState<Config>(initialConfig);
+  const updateTestProviderState = useCallback(
+    (config: Config) => {
+      api.updateTestProviderState(providerId, { config });
+      api.emit(TESTING_MODULE_CONFIG_CHANGE, { providerId, config });
+    },
+    [api, providerId]
+  );
+
+  const [currentConfig, setConfig] = useState<Config>(() => {
+    updateTestProviderState(initialConfig);
+    return initialConfig;
+  });
```

Comment:
This is a pretty wild way to ensure `updateTestProviderState(initialConfig)` only runs on first render. Feels like a `useEffect` with an empty array is more explicit.

[React docs](https://react.dev/reference/react/useState#parameters) says this:

> If you pass a function as initialState, it will be treated as an initializer function. **It should be pure**, should take no arguments, and should return a value of any type.

But this is not pure, it has a side-effect.

Reply:
I experimented with this part of the code a lot. I used `useEffect` initially with an empty array, but the outcome was not the same. Let me try it once again, though!